src/main/ruby/truffleruby/core/posix.rb

```diff
@@ -108,13 +110,28 @@ def self.attach_function_eagerly(native_name, argument_types, return_type,
     if func
       string_args = []
       nfi_args_types = []
-      argument_types.each_with_index do |arg_type, i|
-        if arg_type == :string
-          string_args << i
-          nfi_args_types << '[sint8]'
+      arg_n = 0
+      next_is_varargs = false
+      argument_types = argument_types.dup
+      until argument_types.empty?
+        arg_type = argument_types.shift
+        case arg_type
+        when :"..."
+          raise if argument_types.empty?
+          next_is_varargs = true
+          next
+        when :string
+          string_args << arg_n
+          nfi_arg_type = '[sint8]'
         else
-          nfi_args_types << to_nfi_type(arg_type)
+          nfi_arg_type = to_nfi_type(arg_type)
+        end
+        if next_is_varargs
+          nfi_arg_type = "...#{nfi_arg_type}"
+          next_is_varargs = false
         end
+        nfi_args_types << nfi_arg_type
+        arg_n += 1
```

Comment:
This logic feels quite complicated, and `attach_function_eagerly` needs to be relatively fast in interpreter during startup (although due to this being called on the first time the function is used means it's less critical).

How about specifying the first var arg argument like
`attach_function :fcntl, [:int, :int, [VARARGS, :int]], :int`
or it could also be a helper for clarity:
`attach_function :fcntl, [:int, :int, varargs(:int)], :int`.

And I think we could simply use `def self.vargars(type); :"...#{to_nfi_type type}"; end`, so no code here would need to change.

That doesn't handle multiple varargs, but we have no need for that currently, and the rule could simply be "use varargs() only on the first variadic argument", as a comment on the varargs method.

---

Another approach would be to have truffleposix_{open2,open3} and truffleposix_{fcntl2,fcntl3} defined in LIBTRUFFLEPOSIX, then no need to handle varargs here, which might improve performance, because NFI and libffi must somehow deal with the variable number of arguments vs a fixed number of arguments.

Reply:
> which might improve performance, because NFI and libffi must somehow deal with the variable number of arguments vs a fixed number of arguments.

It's a different code path, but other than that not a perf concern I think.

Reply:
> Another approach would be to have truffleposix_{open2,open3} and truffleposix_{fcntl2,fcntl3} defined in LIBTRUFFLEPOSIX, then no need to handle varargs here, which might improve performance

I don't think there is a performance difference between varargs and not varargs in the NFI or libffi, aside from the intrinsic difference of the calling convention. Afaik on Darwin/AArch64, varargs are always on the stack, not in registers, that might be slower. But there is nothing we can do about that. Even if we can call `truffleposix_open3` with all args in registers, we're putting the argument on the stack one layer down. I doubt that gives us any performance benefit.

Reply:
I think there might be a perf difference simply because e.g. we'd always pass a fixed number of arguments to NFI, and so the Object[] of arguments can be PEA'd since its length would be constant. If we pass sometimes 2 sometimes 3 arguments, and there is no splitting (AFAIK nothing would trigger splitting here) then the Object[] length is not PE constant (unless we compile and inline 2-3 methods or so, might not be the case if the caller is big and not the case in Tier 1), no PEA, and also no argument type profiling. Also NFI will need to serialize those arguments and so either inline caches on the number of arguments, or has conditional/loopy logic to handle extra args.
Maybe that cost is dwarfed by the cost of the native call/libffi though.

Reply:
You need a different signature anyway for the 2 and 3 argument forms. So from the point of view of the NFI, those are completely different code paths. But I don't know how this is implemented on the Ruby side, whether there is a problem from there.

Reply:
The same signature is used for both 2 and 3 arguments here, e.g. open's signature is `([sint8],sint32,...uint32):sint32`.

Reply:
OK, what seems to happen is TruffleRuby always passes 3 arguments (e.g. `fcntl(fd, F_GETFD, 0)`).
So then indeed there is no point to do anything more complicated in this PR, let's just use the `varargs(:int)` solution above.

Reply:
> The same signature is used for both 2 and 3 arguments here, e.g. open's signature is ([sint8],sint32,...uint32):sint32.

No, the signatures would be different in both cases:
* 2 args: `([sint8],sint32):sint32`
* 3 args: `([sint8],sint32,...uint32):sint32.`

Note that the NFI signatures describes the signature of the _call-site_.  For `open` truffleruby always seems to use the 3 args version, but we still have to specify the varargs in the signature.